sklearn/neighbors/tests/test_kde.py

```diff
@@ -151,51 +167,57 @@ def test_kde_pipeline_gridsearch():
-                    scores_const_weight = kde.score_samples(test_points)
-                    sample_const_weight = kde.sample(random_state=1234)
-                    kde.fit(X)
-                    scores_no_weight = kde.score_samples(test_points)
-                    sample_no_weight = kde.sample(random_state=1234)
-                    assert_allclose(scores_const_weight, scores_no_weight)
-                    assert_allclose(sample_const_weight, sample_no_weight)
-
-                    # Test equivalence between sampling and (integer) weights
-                    kde.fit(X, sample_weight=weights)
-                    scores_weight = kde.score_samples(test_points)
-                    sample_weight = kde.sample(random_state=1234)
-                    kde.fit(X_repetitions)
-                    scores_ref_sampling = kde.score_samples(test_points)
-                    sample_ref_sampling = kde.sample(random_state=1234)
-                    assert_allclose(scores_weight, scores_ref_sampling)
-                    assert_allclose(sample_weight, sample_ref_sampling)
-
-                    # Test that sample weights has a non-trivial effect
-                    diff = np.max(np.abs(scores_no_weight - scores_weight))
-                    assert diff > 0.001
-
-                    # Test invariance with respect to arbitrary scaling
-                    scale_factor = rng.rand()
-                    kde.fit(X, sample_weight=(scale_factor * weights))
-                    scores_scaled_weight = kde.score_samples(test_points)
-                    assert_allclose(scores_scaled_weight, scores_weight)
+    rng = np.random.RandomState(0)
+
+    X = rng.rand(n_samples, d)
+    weights = 1 + (10 * X.sum(axis=1)).astype(np.int8)
+    X_repetitions = np.repeat(X, weights, axis=0)
+    n_samples_test = size_test // d
+    test_points = rng.rand(n_samples_test, d)
+
+    kde = KernelDensity(algorithm=algorithm, metric=metric)
+
+    # Test that adding a constant sample weight has no effect
+    kde.fit(X, sample_weight=weights_neutral)
+    scores_const_weight = kde.score_samples(test_points)
+    sample_const_weight = kde.sample(random_state=1234)
+    kde.fit(X)
+    scores_no_weight = kde.score_samples(test_points)
+    sample_no_weight = kde.sample(random_state=1234)
+    assert_allclose(scores_const_weight, scores_no_weight)
+    assert_allclose(sample_const_weight, sample_no_weight)
+
+    # Test equivalence between sampling and (integer) weights
+    kde.fit(X, sample_weight=weights)
+    scores_weight = kde.score_samples(test_points)
+    sample_weight = kde.sample(random_state=1234)
+    kde.fit(X_repetitions)
+    scores_ref_sampling = kde.score_samples(test_points)
+    sample_ref_sampling = kde.sample(random_state=1234)
+    assert_allclose(
+        np.exp(scores_weight), np.exp(scores_ref_sampling), atol=1e-8, rtol=1e-2
+    )
+    assert_allclose(sample_weight, sample_ref_sampling, rtol=1e-2)
```

Comment:
I relaxed the checks here quite a bit, not sure if this is valid. I checked the scipy implementation `gaussian_kde` and it cannot pass the strict version either (unless scipy implementation is also buggy 🤔).

<details><summary>Test code</summary>
<p>

```python
import numpy as np
from scipy.stats import gaussian_kde
from numpy.testing import assert_allclose

n_samples = 400
size_test = 20
rng = np.random.default_rng(0)

X = rng.random((n_samples, 1))
weights = 1 + (10 * X.sum(axis=1)).astype(np.int8)
X_repetitions = np.repeat(X, weights, axis=0)
n_samples_test = size_test
test_points = rng.random((n_samples_test, 1))

kde1 = gaussian_kde(X.T, weights=weights)
scores_weight = kde1.pdf(test_points.T)
sample_weight = kde1.resample(1)
kde2 = gaussian_kde(X_repetitions.T)
scores_ref_sampling = kde2.pdf(test_points.T)
sample_ref_sampling = kde2.resample(1)
assert_allclose(scores_weight, scores_ref_sampling)
assert_allclose(sample_weight, sample_ref_sampling)
```

</p>
</details>

Reply:
I assume that the equivalence is in expectation so the strict check will not work. I think this is something that @snath-xoc might help us here. She is solving a similar issue regarding this equivalence in other estimators. I assume that the KDE could be added to the list of estimator to look after.

Reply:
Thanks for looping me in @glemaitre, @Charlie-XIAO the test itself looks O.K. at a first glance. It may be that due to the stochasticity, the values will only converge to the same in expected but not exact values. I will add this into my tests for these kind of situations and see how it performs.

Reply:
Maybe check that the `bandwidth_` value is the same with integer weights and repetitions before trying to sample from the models.

Reply:
I don't see anything that is expected to behave stochastically at fit time in the code. In particular the `KernelDensity` constructor does not accept a `random_state` argument.

Reply:
`bandwidth_` is not the same if we use `scott` or `silverman` and I don't think they should be the same. If we directly specify a float bandwidth then `bandwidth_` would indeed be the same.